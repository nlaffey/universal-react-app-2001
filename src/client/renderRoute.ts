import * as ReactDOM from 'react-dom';
import { ResolveContext } from '../typings/server';
import { insertCss } from '../utils/css';
import appUniversalRouter from '../universal-router/appUniversalRouter';
import getInitialPropsRecursively, { InitialPropsContext } from '../utils/getInitialPropsRecursively';

declare global {
  // noinspection TsLint
  interface Window {
    initialProps: Object;
  }
}

export interface AppLocation {
  pathname: string;
  search: string;
}

export type RenderRoute = (location: AppLocation, mountingElement: HTMLElement, hydrate?: boolean) => Promise<{}>;
export const renderRoute: RenderRoute =
  async function renderRoute(location: AppLocation, mountingElement: HTMLElement, hydrate?: boolean) {
    return new Promise(async (resolve) => {
      const resolveObject: ResolveContext = {
        pathname: location.pathname,
        query: location.search,
        context: { insertCss }
      };
      const component = await appUniversalRouter.resolve(resolveObject);
      const initialPropsContext: InitialPropsContext = { resolveContext: resolveObject, port: null };
      const initialProps = await getInitialPropsRecursively(component, initialPropsContext);
      const resolveObjectWithProps = { ...resolveObject, initialProps };
      const componentWithProps = await appUniversalRouter.resolve(resolveObjectWithProps);
      if (hydrate) {
        resolve(ReactDOM.hydrate(componentWithProps, mountingElement));
      } else {
        resolve(ReactDOM.render(componentWithProps, mountingElement));
      }
    });
  };
