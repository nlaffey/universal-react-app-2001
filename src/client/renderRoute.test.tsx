import { AppLocation, RenderRoute } from './renderRoute';

describe('renderRoute', () => {

  let renderRoute: RenderRoute;
  beforeAll(() => {
    jest.mock('../universal-router/appUniversalRouter');
    jest.mock('../utils/getInitialPropsRecursively');
    jest.mock('react-dom', () => ({
      hydrate: jest.fn(),
      render: jest.fn()
    }));
    renderRoute = require('./renderRoute').renderRoute;
  });
  test('should call getInitialPropsRecursively', async () => {
    const location: AppLocation = {
      pathname: 'mock/path/name',
      search: '?mockSearch'
    };
    const mockElement: any = { mockElement: true };
    await renderRoute(location, mockElement, true);
    const getInitialPropsRecursively = require('../utils/getInitialPropsRecursively').default;
    expect(getInitialPropsRecursively.mock.calls).toMatchSnapshot();
  });

  test('should call resolve twice, once without props, once with props', async () => {
    const location: AppLocation = {
      pathname: 'mock/path/name',
      search: '?mockSearch'
    };
    const mockElement: any = { mockElement: true };
    await renderRoute(location, mockElement, true);
    const { resolve } = require('../universal-router/appUniversalRouter').default;
    expect(resolve.mock.calls).toMatchSnapshot();
  });

  test('should call hydrate when is initial render', async () => {
    const location: AppLocation = {
      pathname: 'mock/path/name',
      search: '?mockSearch'
    };
    const mockElement: any = { mockElement: true };
    await renderRoute(location, mockElement, true);
    const { hydrate } = require('react-dom');
    expect(hydrate.mock.calls).toMatchSnapshot();
  });

  test('should call render when not initial render', async () => {
    const location: AppLocation = {
      pathname: 'mock/path/name',
      search: '?mockSearch'
    };
    const mockElement: any = { mockElement: true };
    await renderRoute(location, mockElement, false);
    const { render } = require('react-dom');
    expect(render.mock.calls).toMatchSnapshot();
  });
});
