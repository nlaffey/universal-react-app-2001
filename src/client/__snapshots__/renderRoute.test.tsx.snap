// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderRoute should call getInitialPropsRecursively 1`] = `
Array [
  Array [
    Object {
      "mockComponent": true,
    },
    Object {
      "port": null,
      "resolveContext": Object {
        "context": Object {
          "insertCss": [Function],
        },
        "pathname": "mock/path/name",
        "query": "?mockSearch",
      },
    },
  ],
]
`;

exports[`renderRoute should call hydrate when is initial render 1`] = `
Array [
  Array [
    Object {
      "mockComponent": true,
    },
    Object {
      "mockElement": true,
    },
  ],
  Array [
    Object {
      "mockComponent": true,
    },
    Object {
      "mockElement": true,
    },
  ],
  Array [
    Object {
      "mockComponent": true,
    },
    Object {
      "mockElement": true,
    },
  ],
]
`;

exports[`renderRoute should call render when not initial render 1`] = `
Array [
  Array [
    Object {
      "mockComponent": true,
    },
    Object {
      "mockElement": true,
    },
  ],
]
`;

exports[`renderRoute should call resolve twice, once without props, once with props 1`] = `
Array [
  Array [
    Object {
      "context": Object {
        "insertCss": [Function],
      },
      "pathname": "mock/path/name",
      "query": "?mockSearch",
    },
  ],
  Array [
    Object {
      "context": Object {
        "insertCss": [Function],
      },
      "initialProps": Object {
        "mockInitialProps": true,
      },
      "pathname": "mock/path/name",
      "query": "?mockSearch",
    },
  ],
  Array [
    Object {
      "context": Object {
        "insertCss": [Function],
      },
      "pathname": "mock/path/name",
      "query": "?mockSearch",
    },
  ],
  Array [
    Object {
      "context": Object {
        "insertCss": [Function],
      },
      "initialProps": Object {
        "mockInitialProps": true,
      },
      "pathname": "mock/path/name",
      "query": "?mockSearch",
    },
  ],
]
`;
